syz-verifier/verifier.go

```diff
@@ -31,24 +36,180 @@ type Verifier struct {
 	// - <workdir>/corpus.db: corpus with interesting programs
 	// - <workdir>/<OS-Arch>/instance-x: per VM instance temporary files
 	// grouped by OS/Arch
-	workdir       string
-	crashdir      string
-	resultsdir    string
-	target        *prog.Target
-	runnerBin     string
-	executorBin   string
-	choiceTable   *prog.ChoiceTable
-	rnd           *rand.Rand
-	progIdx       int
-	addr          string
-	srv           *RPCServer
-	calls         map[*prog.Syscall]bool
-	reasons       map[*prog.Syscall]string
-	reportReasons bool
-	stats         *Stats
-	statsWrite    io.Writer
-	newEnv        bool
-	reruns        int
+	workdir           string
+	crashdir          string
+	resultsdir        string
+	target            *prog.Target
+	runnerBin         string
+	executorBin       string
+	progGeneratorInit sync.WaitGroup
+	choiceTable       *prog.ChoiceTable
+	rnd               *rand.Rand
+	progIdx           int
+	addr              string
+	srv               *RPCServer
+	calls             map[*prog.Syscall]bool
+	reasons           map[*prog.Syscall]string
+	reportReasons     bool
+	stats             *Stats
+	statsWrite        io.Writer
+	newEnv            bool
+	reruns            int
+
+	// We use single queue for every kernel environment.
+	tasksMutex     sync.Mutex
+	onTaskAdded    *sync.Cond
+	kernelEnvTasks [][]*ExecTaskQueue
+}
+
+func (vrf *Verifier) Init() {
+	vrf.progGeneratorInit.Add(1)
+
+	vrf.onTaskAdded = sync.NewCond(&vrf.tasksMutex)
+
+	vrf.kernelEnvTasks = make([][]*ExecTaskQueue, len(vrf.pools))
+	for i := range vrf.kernelEnvTasks {
+		vrf.kernelEnvTasks[i] = make([]*ExecTaskQueue, EnvironmentsCount)
+	}
+
+	for i := range vrf.kernelEnvTasks {
+		for j := range vrf.kernelEnvTasks[i] {
+			vrf.kernelEnvTasks[i][j] = MakeExecTaskQueue()
+		}
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+	vrf.srv = srv
+}
+
+func (vrf *Verifier) StartProgramsAnalysis() {
+	go func() {
+		vrf.progGeneratorInit.Wait()
+
+		type AnalysisResult struct {
+			Diff []*ExecResult
+			Prog *prog.Prog
+		}
+
+		results := make(chan *AnalysisResult)
+		go func() {
+			for result := range results {
+				if result.Diff != nil {
+					vrf.SaveDiffResults(result.Diff, result.Prog)
+				}
+			}
+		}()
+
+		sem := semaphore.NewWeighted(int64(100))
+		ctx := context.Background()
+		for {
+			sem.Acquire(ctx, 1)
+			go func(prog *prog.Prog) {
+				defer sem.Release(1)
+				results <- &AnalysisResult{
+					vrf.TestProgram(prog),
+					prog,
+				}
+			}(vrf.generate())
+		}
+	}()
+}
+
+func (vrf *Verifier) GetRunnerTask(kernel int, existing EnvDescr) *rpctype.ExecTask {
+	vrf.tasksMutex.Lock()
+	defer vrf.tasksMutex.Unlock()
+
+	for {
+		for env := existing; env >= AnyEnvironment; env-- {
+			if task, ok := vrf.kernelEnvTasks[kernel][env].PopTask(); ok {
+				return task.ToRPC()
+			}
+		}
+
+		vrf.onTaskAdded.Wait()
+	}
+}
+
+func PutExecResult(result *ExecResult) {
+	go func() {
```

Comment:
Why do we need this in a separate goroutine?

Reply:
Premature optimization attempt. Wanted to return control as quick as possible.
Remove.

Reply:
Goroutines are slower and introduce lots of overhead as compared to direct control flow.